src/pip/_internal/index.py

```diff
@@ -226,6 +226,119 @@ def _get_html_page(link, session=None):
         return HTMLPage(resp.content, resp.url, resp.headers)
 
 
+class _LinkParseError(ValueError):
+    pass
+
+
+class _LinkNotAFile(_LinkParseError):
+    pass
+
+
+class _LinkFormatUnsupported(_LinkParseError):
+    pass
+
+
+class _BinaryLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _MacOSXZip(_LinkParseError):
+    pass
+
+
+class _IncorrectWheelName(_LinkParseError):
+    pass
+
+
+class _IncompatibleWheel(_LinkParseError):
+    pass
+
+
+class _SourceLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _PythonVersionNotSupported(_LinkParseError):
+    pass
+
+
+def _egg_info_matches(egg_info, canonical_name):
+    """Pull the version part out of a string.
+
+    :param egg_info: The string to parse. E.g. foo-2.1
+    :param canonical_name: The canonicalized name of the package this
+        belongs to.
+    """
+    try:
+        version_start = _find_name_version_sep(egg_info, canonical_name) + 1
+    except ValueError:
+        return None
+    version = egg_info[version_start:]
+    if not version:
+        return None
+    return version
+
+
+_PYTHON_VERSION_RE = re.compile(r'-py([123]\.?[0-9]?)$')
+
+
+def _parse_package_version_non_wheel(egg_info, search):
```

Comment:
A docstring would be helpful here -- in particular what type do the arguments have?